### ISSUE TITLE
Incorrect Type for Custom `@property` Attribute

### ISSUE THREAD

**Issue by YannickAlex07** [microsoft/pyright#8027]
**Describe the bug**

Hey, I have encountered an issue with an annotation in the `pyspark` package (version `3.5.1`).

Apparently they are doing some shenanigans and created a custom `@classproperty` decorator that inherits from the `@property` decorator. [Here](https://github.com/apache/spark/blob/69afd4be9c93cb31a840b969ed1984c0b6b92f8e/python/pyspark/sql/session.py#L148-L178) is the code of the `classproperty`-class. Note that the `__get__` attribute is hinted to return a `SparkSession.Builder` instance. They then declare a property [`builder`](https://github.com/apache/spark/blob/69afd4be9c93cb31a840b969ed1984c0b6b92f8e/python/pyspark/sql/session.py#L608) on the actual `SparkSession` class that uses this custom decorator (weirdly as a function here) and is also hinted to return a `Builder` instance.

However now when I want to use the `builder` attribute, `pyright` infers the type to be `classproperty` instead of `Builder`. On hover it shows the following: `(property) builder: classproperty`.

_Note: They seemed to change the `builder` property between `3.5.1` and the current master version. I would guess that it shouldn't make a big different but for me it is currently locally declared as an actual method with the annotation:_

```python
@classproperty
def builder(cls) -> Builder:
    """
    ...
    """
    return cls.Builder()
``` 

**Code or Screenshots**

The following code will already show errors:

```python
from pyspark.sql import SparkSession

spark = SparkSession.builder.appName("Test").getOrCreate()
```

It will complain about the `.appName(...)` part with the following error:

`Cannot access attribute "appName" for class "classproperty" Attribute "appName" is unknown`

**VS Code extension or command-line**

Python: `3.11.6`
PySpark: `3.5.1`
VS Code Extension (Pylance): `v2024.5.103 (pre-release)`

**Comment by erictraut** [comment]
`property` requires significant special casing in a static type checker. That's because `property` does a bunch of magic behind the scenes with its `__fget__`, `__fset__` and `__fdel__` attributes. The `property` class is also not defined as generic, but it acts like a generic.

Because of this special casing, attempts to subclass `property` will not work for static typing because those special-case behaviors can't be applied.

If the pyspark maintainers want their code to work with static typing, they should avoid subclassing `property` and instead create a class that implements the descriptor protocol itself, without relying on the "magic" performed by `property`.

I don't see anything actionable here for pyright, so I'm going to close the issue.

**Comment by YannickAlex07** [comment]
Ah that explains it, thanks you for the explanation! I will open something in the PySpark issue tracker then, even though I guess they plan to change this anyway once Python 3.8 support is dropped in October.
